Update client tooling.

peterhuene opened the pull request:
With these changes, `cargo-component` can now remove its own representation of storing registry data and (soon) share the same data on disk as `warg-cli`.

`warg-client` changes:

* Fully document the `warg-client` crate.
* Implement lock files for client storage directories.
* Simplify the client implementation.
* Enforce HTTPS for non-local registry servers.
* Add error response handling to API client.
* Put package name from publishing request body.
* Added tracing to the client.

`warg-server` changes:

* The publishing service now accepts the package name in the request body, removing the need for encoding it in the URL path.
* A trace layer has been added to the root router enabling the server to print out information for when a request starts and finishes.

`warg-cli` changes:

* A tracing subscriber was added to trace client messages.
* The `install` command was renamed to `download`.
* The `info` command was updated to display more information.
* A `publish init` command was added for publishing an initial record for a package (pending or immediately).
* The `publish release` command can now publish immediately if there is no pending publish.

Dependencies have also been updated to latest and the README was expanded to include information on how to run the server and client.

Kylebrown9 commented:
Can you demonstrate the usage of the new cli commands in the README?

peterhuene commented:
@Kylebrown9 if you don't mind holding off on the review for just a little bit, I think I might rip out the "local" storage implementation here after all. Vendoring will be done differently in cargo-component.

It should simplify these changes a little.

peterhuene commented:
I'll update the README as part of this PR.

Kylebrown9 commented on src/commands/publish.rs:
```diff
@@ -42,82 +49,104 @@ impl PublishCommand {
     pub init: bool,
+    /// Queue the release for publishing with the `publish submit` command.
+    #[clap(long)]
+    pub queue: bool,
```
The proposed `--queue` flag would have to be added to every future command that can be deferred (e.g. yank, grant, revoke). I'm not sure I like the ergonomics of this.

peterhuene replied:
After testing the DX a bit, I decided to just put back the `publish start` command, but also add a `publish init` command that either enqueues an init record (for a pending publish) or immediately tries to initialize the package if there is no pending publish, similar to how `publish release` now operates.

peterhuene commented:
Pushed a rebase on main (haven't addressed review feedback yet) as I just wanted to see how this PR does on a Windows CI given the lock file implementation.

peterhuene commented on crates/client/src/lib.rs:
```diff
@@ -1,328 +1,560 @@
+                    .ok_or(ClientError::OperationNotSupported)?;
+
+                let mut info = PackageInfo::new(name);
+                self.update_checkpoint(&api.latest_checkpoint().await?, [&mut info])
```
This is primarily for fetching a missing package log when downloading a requested version, so it bounds the update to only the requested package.

peterhuene commented on crates/server/src/api/content.rs:
```diff
@@ -23,25 +25,31 @@ pub struct ContentConfig {
         Ok(Router::new()
+            .route("/:content_id", get(get_content))
             .route("/", post(upload_content))
-            .fallback_service(get_service(ServeDir::new(&self.content_path)).handle_error(
```
As discussed 1:1, this was an attempt to make the content URLs be more consistent with other services that use `DynHash`-based ids in their paths so that the client doesn't have to know to replace the `:` with `-`.

I will back out this change and we'll potentially solve the inconsistency another way in the future.

peterhuene commented on crates/server/src/services/core.rs:
```diff
@@ -113,6 +113,7 @@ pub struct CoreService {
 #[derive(Debug)]
 enum Message {
     SubmitPackageRecord {
+        package_id: LogId,
```
Agreed, this is redundant. As we discussed, the right change is to simply not have the log id (or name like it is done now) in the publishing endpoint at all and just pass the name in the body, allowing the service to derive the id.

I'll make that change.

peterhuene commented on src/commands.rs:
```diff
@@ -31,9 +31,16 @@ pub struct CommonOptions {
 impl CommonOptions {
+    /// Locks the client storage.
+    pub fn lock_storage(&self) -> Result<FileSystemStorage> {
+        FileSystemStorage::lock(&self.storage, |path| {
```
As discussed, the point was as a cheap mechanism to allow library users to print a message just prior to the blocking operation.

This will be changed to expose a `try_lock` implementation that can be called first, allowing consumers to print a message if desired (or just bail the operation), and then proceed to call `lock`.

peterhuene commented on src/commands/publish.rs:
```diff
@@ -42,82 +49,104 @@ impl PublishCommand {
+        let mut info = storage
+            .load_publish_info()
+            .await?
+            .unwrap_or_else(|| PublishInfo {
```
Agreed, this should be solved with when I fix this with implicit batching.

peterhuene commented:
@Kylebrown9 okay, I think I pushed everything we discussed, plus the removal of "local" registry information in client storage.

Kylebrown9 commented on README.md:
````diff
@@ -44,7 +44,116 @@ updated to install the crates.io package once a proper release is made.
+
+```
+warg-cli publish start hello
+warg-cli publish init hello
````
We should eventually allow users to imply the name if they're already in a batch, but we don't have to do that in this pr.

Kylebrown9 reviewed (approved):
Thanks for the changes!
